gibbon-compiler/src/Gibbon/Common.hs

```diff
@@ -126,27 +132,64 @@ cleanFunName f =
           else '_'
         | c <- fromVar f ]
 
+
 toEndV :: Var -> Var
 toEndV = varAppend "end_"
 
+toEndVRegVar :: RegVar -> RegVar 
+toEndVRegVar (SingleR v) = SingleR (toEndV v)
+toEndVRegVar (SoARv regvar fieldRegs) = SoARv (toEndVRegVar regvar) (L.map (\(k, freg) -> (k, toEndVRegVar freg)) fieldRegs)
+
 toEndVLoc :: LocVar -> LocVar 
 toEndVLoc loc = case loc of 
                     Single v -> Single (toEndV v)
+                    SoA dcon fieldLocs -> SoA (toEndV dcon) (L.map (\(k, floc) -> (k, toEndVLoc floc)) fieldLocs)
 
 toSeqV :: Var -> Var
 toSeqV v = varAppend v (toVar "_seq")
 
 -- | A location variable stores the abstract location. 
 type Location = Var
+-- | The position or index of a field in a data constructor value.
+type FieldIndex = Int 
+-- | A data constructor is a String type in the compiler
+type DataCon = String
+
+-- | Single: For storing a single location, useful for adding a cursor in a region. 
+-- | SoA: A location signature for a structure of arrays representation. 
+--        The first location points to a location in the data constructor buffer. 
+--        The list includes locations for each field and a tuple storing 
+--        information about which data constructor and corresponding index the field 
+--        comes from. 
+-- TODO: I think the type for an SoA location is not right here 
+-- A Location should work for a data constructor buffer 
+-- However, imagine if we have a data type definition of 
+-- data Foo = A Int List Tree Tree | Leaf
+-- here the List would be in its own buffer potentially
+-- so we have nesting of SoA locations 
+-- Possibly need to change Location in SoA to LocVar, a recursive data type 
+-- But for simple data types like data List = Cons Int List | Nil 
+-- this should work just fine. 
+-- One reason I don't want to make an SoA location recursive is that you might 
+-- want to make the level of factoring limited to only depth = 1
+-- more factoring than a depth of level one might slow down too much
+-- data List2 = Cons2 Int List List2 | Nil2
+-- In the SoA representation, it is guaranteed that the data constructors should 
+-- Remain in the same buffer. Hence, its fine to have its type as Location instead 
+-- of LocVar.
+
+data LocVar = Single Location | SoA Location [((DataCon, FieldIndex), LocVar)]
```

Comment:
Can you, please, move comments about individual constructors from above to the definition of constructors themselves? See this for an example of how constructors are annotated with Haddock comments: https://hackage-content.haskell.org/package/cabal-install-3.14.2.0/docs/src/Distribution.Client.TargetSelector.html#SubComponentTarget